File: #144_Binary_Tree_Preorder_Traversal/Solution.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
    	List<Integer> result = new LinkedList();
		if (root == null) return result;
		Deque<TreeNode> stack = new LinkedList();
		while (root != null || !stack.isEmpty()) {
			if (root != null) {
				result.add(root.val);
				stack.push(root);
				root = root.left;
			}
			else {
				root = stack.pop();
				root = root.right;
			}
		}
		return result;
    }
}

File: #144_Binary_Tree_Preorder_Traversal/Solution.py
'''
Recursive solution is trivial
'''

# Definition for a binary tree node.
class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:

	def __init__(self):
		self.ans = None

	# @param {TreeNode} root
	# @return {integer[]}
	def preorderTraversal(self, root):
		self.ans = []
		self.preorder(root)
		return self.ans

	def preorder(self, root):
		if not root:
			return
		self.ans.append(root.val)
		self.preorder(root.left)
		self.preorder(root.right)